pkg/operator/controller/service-dns/controller.go

```diff
@@ -0,0 +1,256 @@
+		}
+	}
+	return !reflect.DeepEqual(x, y)
+}
+
+// Config holds all the configuration that must be provided when creating the
+// controller.
+type Config struct {
+	// OperandNamespace is the namespace in which to watch for services and
+	// dnsrecords and in which to create dnsrecords.
+	OperandNamespace string
+}
+
+// reconciler handles the actual service reconciliation logic.
+type reconciler struct {
+	config Config
+
+	client   client.Client
+	cache    cache.Cache
+	recorder record.EventRecorder
+}
+
+// Reconcile expects request to refer to a service and creates or reconciles a
+// dnsrecord.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	var service corev1.Service
+	if err := r.cache.Get(ctx, request.NamespacedName, &service); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("service not found; reconciliation will be skipped", "request", request)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if len(service.Spec.Selector[gatewayNameLabelKey]) == 0 {
+		log.Info(`service selector has no "`+gatewayNameLabelKey+`" label; reconciliation will be skipped`, "request", request)
+		return reconcile.Result{}, nil
+	}
+
+	var gateway gatewayapiv1beta1.Gateway
+	gatewayName := types.NamespacedName{
+		Namespace: service.Namespace,
+		Name:      service.Spec.Selector[gatewayNameLabelKey],
+	}
+	if err := r.cache.Get(ctx, gatewayName, &gateway); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("gateway not found; reconciliation will be skipped", "request", request)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	domains := getGatewayHostnames(&gateway)
+	var errs []error
+	errs = append(errs, r.ensureDNSRecordsForGateway(ctx, &gateway, &service, domains.List())...)
+	errs = append(errs, r.deleteStaleDNSRecordsForGateway(ctx, &gateway, &service, domains)...)
+	return reconcile.Result{}, utilerrors.NewAggregate(errs)
+}
+
+// getGatewayHostnames returns a sets.String with the hostnames from the given
+// gateway's listeners.
+func getGatewayHostnames(gateway *gatewayapiv1beta1.Gateway) sets.String {
+	domains := sets.NewString()
+	for _, listener := range gateway.Spec.Listeners {
+		if listener.Hostname == nil || len(*listener.Hostname) == 0 {
+			continue
+		}
+		domains.Insert(string(*listener.Hostname))
+	}
+	return domains
+}
+
+// ensureDNSRecordsForGateway ensures that a DNSRecord CR exists, associated
+// with the given gateway and service, for each of the given domains.  It
+// returns a list of any errors that result from ensuring those DNSRecord CRs.
+func (r *reconciler) ensureDNSRecordsForGateway(ctx context.Context, gateway *gatewayapiv1beta1.Gateway, service *corev1.Service, domains []string) []error {
+	labels := map[string]string{
+		gatewayNameLabelKey: gateway.Name,
+	}
+	for k, v := range service.Labels {
```

Comment:
What's the motivation behind taking all of the service labels and placing them on the dns record?

Reply:
I figured the service labels likely have useful information for understanding how these resources relate to a given gateway or manager/controller.  I can drop it if it isn't useful.

Reply:
If it doesn't hurt anything, I don't mind. Just was curious, maybe it would help us in the future.

Reply:
Come to think of it, I might have been inspired by Istio:

> Annotations and labels on the `Gateway` will be copied to the `Service` and `Deployment`.

https://istio.io/latest/docs/tasks/traffic-management/ingress/gateway-api/#automated-deployment